pdns/recursordist/rec-eventtrace.hh

```diff
@@ -0,0 +1,248 @@
+  static std::string toString(const Value_t& v)
+  {
+    if (std::holds_alternative<std::nullopt_t>(v)) {
+      return "";
+    }
+    else if (std::holds_alternative<bool>(v)) {
+      return std::to_string(std::get<bool>(v));
+    }
+    else if (std::holds_alternative<int64_t>(v)) {
+      return std::to_string(std::get<int64_t>(v));
+    }
+    else if (std::holds_alternative<std::string>(v)) {
+      return std::get<std::string>(v);
+    }
+    else if (std::holds_alternative<PacketBuffer>(v)) {
+      const PacketBuffer& p = std::get<PacketBuffer>(v);
+      return makeHexDump(std::string(reinterpret_cast<const char*>(p.data()), p.size()));
+    }
+    return "?";
+  }
+
+  struct Entry
+  {
+    Entry(Value_t& v, EventType e, bool start, int64_t ts) :
+      d_value(v), d_ts(ts), d_event(e), d_start(start)
+    {
+    }
+    Entry(Value_t& v, const std::string& custom, bool start, int64_t ts) :
+      d_value(v), d_custom(custom), d_ts(ts), d_event(CustomEvent), d_start(start)
+    {
+    }
+    Value_t d_value;
+    std::string d_custom;
+    int64_t d_ts;
+    EventType d_event;
+    bool d_start;
+
+    std::string toString() const
+    {
+      std::string v = RecEventTrace::toString(d_value);
+      if (!v.empty()) {
+        v = "," + v;
+      }
+      std::string name = RecEventTrace::toString(d_event);
+      if (d_event == EventType::CustomEvent) {
+        name += ":" + d_custom;
+      }
+
+      return name + "(" + std::to_string(d_ts) + v + (d_start ? ")" : ",done)");
+    }
+  };
+
+  void setEnabled(bool flag)
+  {
+    assert(d_status != Invalid);
+    d_status = flag ? Enabled : Disabled;
+  }
+
+  bool enabled() const
+  {
+    return d_status == Enabled;
+  }
+
+  template <class E>
+  void add(E e, Value_t v, bool start)
+  {
+    assert(d_status != Invalid);
+    if (d_status == Disabled) {
+      return;
+    }
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    int64_t stamp = ts.tv_nsec + ts.tv_sec * 1000000000;
+    stamp -= d_base;
+    d_events.emplace_back(v, e, start, stamp);
+  }
+
+  template <class E>
+  void add(E e)
+  {
+    add(e, Value_t(std::nullopt), true);
+  }
+
+  // We store uint32 in an int64_t
+  template <class E>
+  void add(E e, uint32_t v, bool start)
+  {
+    add(e, static_cast<int64_t>(v), start);
+  }
+  // We store int32 in an int64_t
+  template <class E>
+  void add(E e, int32_t v, bool start)
+  {
+    add(e, static_cast<int64_t>(v), start);
+  }
+
+  template <class E, class T>
+  void add(E e, T v, bool start)
+  {
+    add(e, Value_t(v), start);
+  }
+
+  void clear()
+  {
+    d_events.clear();
+    reset();
+  }
+  void reset()
+  {
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    d_base = ts.tv_nsec + ts.tv_sec * 1000000000;
+    d_status = Disabled;
+  }
+
+  std::string toString() const
+  {
+    assert(d_status != Invalid);
+    if (d_status == Disabled) {
+      return "Disabled\n";
+    }
+    std::string ret = "eventTrace [";
```

Comment:
I wonder if using an `ostringstream` would make this easier, but that does not really matter.